File: packages/shared-ts/src/validation.ts
import type {
  GameConfig,
  PaylineDefinition,
  PayoutRuleDefinition,
  PayoutSymbolDefinition,
  SymbolDefinition,
} from './contracts.js';
import { DomainValidationError } from './errors.js';
import {
  MAX_SIGNATURE_CELL_WIDTH,
  MAX_SYMBOL_MOBILE_CODE,
} from './signature.js';

function requireNonEmpty(
  value: string,
  code: 'invalid_game' | 'invalid_payline_id' | 'invalid_symbol',
): void {
  if (value.trim().length === 0) {
    throw new DomainValidationError(code, 'Value must not be empty.');
  }
}

function requireNonNegativeInteger(
  value: number,
  code: 'invalid_payout' | 'invalid_spin_cost',
  label: string,
): void {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new DomainValidationError(
      code,
      `${label} must be a non-negative safe integer.`,
    );
  }
}

export function validateBoardDimensions(rows: number, columns: number): void {
  if (
    !Number.isSafeInteger(rows) ||
    !Number.isSafeInteger(columns) ||
    rows < 1 ||
    columns < 1
  ) {
    throw new DomainValidationError(
      'invalid_dimensions',
      'Board dimensions must be positive safe integers.',
    );
  }
}

function validateSymbols(
  symbols: readonly SymbolDefinition[],
  cellWidth: number,
): void {
  if (symbols.length === 0) {
    throw new DomainValidationError(
      'invalid_symbol',
      'Game must define at least one symbol.',
    );
  }

  const mobileCodes = new Set<number>();
  const codes = new Set<string>();
  const maximumCode = Math.min(10 ** cellWidth - 1, MAX_SYMBOL_MOBILE_CODE);

  for (const symbol of symbols) {
    requireNonEmpty(symbol.code, 'invalid_symbol');
    requireNonEmpty(symbol.name, 'invalid_symbol');

    if (
      !Number.isSafeInteger(symbol.mobileCode) ||
      symbol.mobileCode < 1 ||
      symbol.mobileCode > maximumCode
    ) {
      throw new DomainValidationError(
        'invalid_symbol_code',
        `Symbol mobile code must fit signature width ${cellWidth}.`,
      );
    }
    if (!Number.isSafeInteger(symbol.displayOrder) || symbol.displayOrder < 0) {
      throw new DomainValidationError(
        'invalid_symbol',
        'Symbol display order must be a non-negative integer.',
      );
    }
    if (mobileCodes.has(symbol.mobileCode)) {
      throw new DomainValidationError(
        'duplicate_symbol_mobile_code',
        `Duplicate symbol mobile code ${symbol.mobileCode}.`,
      );
    }
    if (codes.has(symbol.code)) {
      throw new DomainValidationError(
        'duplicate_symbol_code',
        `Duplicate symbol code ${symbol.code}.`,
      );
    }
    mobileCodes.add(symbol.mobileCode);
    codes.add(symbol.code);
  }
}

export function validateGameConfig(game: GameConfig): void {
  requireNonEmpty(game.id, 'invalid_game');
  requireNonEmpty(game.code, 'invalid_game');
  requireNonEmpty(game.name, 'invalid_game');
  validateBoardDimensions(game.rows, game.columns);
  requireNonNegativeInteger(game.spinCost, 'invalid_spin_cost', 'Spin cost');

  if (
    !Number.isSafeInteger(game.signatureCellWidth) ||
    game.signatureCellWidth < 1 ||
    game.signatureCellWidth > MAX_SIGNATURE_CELL_WIDTH
  ) {
    throw new DomainValidationError(
      'invalid_cell_width',
      `Signature cell width must be between 1 and ${MAX_SIGNATURE_CELL_WIDTH}.`,
    );
  }

  validateSymbols(game.symbols, game.signatureCellWidth);
}

function allowedSymbolCodes(game: GameConfig): ReadonlySet<number> {
  return new Set(game.symbols.map((symbol) => symbol.mobileCode));
}

function validatePopulatedCell(
  cell: number,
  allowedCodes: ReadonlySet<number>,
): void {
  if (!Number.isSafeInteger(cell) || !allowedCodes.has(cell)) {
    throw new DomainValidationError(
      'invalid_board_symbol',
      `Symbol mobile code ${cell} does not belong to the game.`,
    );
  }
}

export function validateFullBoard(
  cells: readonly number[],
  game: GameConfig,
): void {
  validateGameConfig(game);
  const expectedLength = game.rows * game.columns;
  if (cells.length !== expectedLength) {
    throw new DomainValidationError(
      'invalid_board_length',
      `Board contains ${cells.length} cells; expected ${expectedLength}.`,
    );
  }

  const allowedCodes = allowedSymbolCodes(game);
  cells.forEach((cell) => validatePopulatedCell(cell, allowedCodes));
}

export function validateLayoutBoard(
  cells: readonly number[],
  game: GameConfig,
): void {
  validateGameConfig(game);
  const expectedLength = game.rows * game.columns;
  if (cells.length !== expectedLength) {
    throw new DomainValidationError(
      'invalid_board_length',
      `Board contains ${cells.length} cells; expected ${expectedLength}.`,
    );
  }
  const allowedCodes = new Set([0, ...allowedSymbolCodes(game)]);
  cells.forEach((cell) => validatePopulatedCell(cell, allowedCodes));
}

export function validateBoardPrefix(
  cells: readonly (number | null)[],
  game: GameConfig,
): void {
  validateGameConfig(game);
  const expectedLength = game.rows * game.columns;
  if (cells.length !== expectedLength) {
    throw new DomainValidationError(
      'invalid_board_length',
      `Board contains ${cells.length} cells; expected ${expectedLength}.`,
    );
  }

  const allowedCodes = allowedSymbolCodes(game);
  let reachedEmptyCell = false;
  for (const cell of cells) {
    if (cell === null) {
      reachedEmptyCell = true;
      continue;
    }
    if (reachedEmptyCell) {
      throw new DomainValidationError(
        'non_prefix_board',
        'A populated cell cannot occur after an empty prefix cell.',
      );
    }
    validatePopulatedCell(cell, allowedCodes);
  }
}

export function validateRowPath(
  rowPath: readonly number[],
  rows: number,
  columns: number,
): void {
  validateBoardDimensions(rows, columns);
  if (rowPath.length !== columns) {
    throw new DomainValidationError(
      'invalid_row_path_length',
      `Payline contains ${rowPath.length} rows; expected ${columns}.`,
    );
  }
  if (
    rowPath.some(
      (rowIndex) =>
        !Number.isSafeInteger(rowIndex) || rowIndex < 0 || rowIndex >= rows,
    )
  ) {
    throw new DomainValidationError(
      'invalid_row_index',
      `Every payline row index must be between 0 and ${rows - 1}.`,
    );
  }
}

export function validatePaylines(
  paylines: readonly PaylineDefinition[],
  game: GameConfig,
): void {
  validateGameConfig(game);
  const paths = new Set<string>();
  const ids = new Set<string>();

  for (const payline of paylines) {
    requireNonEmpty(payline.id, 'invalid_payline_id');
    validateRowPath(payline.rowPath, game.rows, game.columns);
    const pathKey = payline.rowPath.join(',');
    if (paths.has(pathKey) || ids.has(payline.id)) {
      throw new DomainValidationError(
        'duplicate_payline',
        'Payline id and row path must be unique.',
      );
    }
    paths.add(pathKey);
    ids.add(payline.id);
  }
}

export function validatePayoutSymbols(
  payoutSymbols: readonly PayoutSymbolDefinition[],
  game: GameConfig,
): void {
  validateGameConfig(game);
  const symbolsByCode = new Map(
    game.symbols.map((symbol) => [symbol.mobileCode, symbol]),
  );
  const configuredSymbols = new Set<number>();

  for (const payoutSymbol of payoutSymbols) {
    const symbol = symbolsByCode.get(payoutSymbol.symbolMobileCode);
    if (symbol === undefined) {
      throw new DomainValidationError(
        'invalid_board_symbol',
        `Payout symbol ${payoutSymbol.symbolMobileCode} does not belong to the game.`,
      );
    }
    if (symbol.isWildcard) {
      throw new DomainValidationError(
        'wildcard_payout_symbol',
        'Wildcard symbols cannot define payout configuration.',
      );
    }
    if (
      !Number.isSafeInteger(payoutSymbol.minimumMatchLength) ||
      payoutSymbol.minimumMatchLength < 2 ||
      payoutSymbol.minimumMatchLength > game.columns
    ) {
      throw new DomainValidationError(
        'invalid_minimum_match_length',
        `Minimum match length must be between 2 and ${game.columns}.`,
      );
    }
    if (configuredSymbols.has(payoutSymbol.symbolMobileCode)) {
      throw new DomainValidationError(
        'duplicate_payout_symbol',
        `Duplicate payout configuration for symbol ${payoutSymbol.symbolMobileCode}.`,
      );
    }
    configuredSymbols.add(payoutSymbol.symbolMobileCode);
  }
}

export function validatePayoutRules(
  rules: readonly PayoutRuleDefinition[],
  payoutSymbols: readonly PayoutSymbolDefinition[],
  game: GameConfig,
): void {
  validatePayoutSymbols(payoutSymbols, game);
  const symbolsByCode = new Map(
    game.symbols.map((symbol) => [symbol.mobileCode, symbol]),
  );
  const payoutSymbolsByCode = new Map(
    payoutSymbols.map((symbol) => [symbol.symbolMobileCode, symbol]),
  );
  const keys = new Set<string>();

  for (const rule of rules) {
    const symbol = symbolsByCode.get(rule.symbolMobileCode);
    if (symbol === undefined) {
      throw new DomainValidationError(
        'invalid_board_symbol',
        `Payout symbol ${rule.symbolMobileCode} does not belong to the game.`,
      );
    }
    if (symbol.isWildcard) {
      throw new DomainValidationError(
        'wildcard_payout_rule',
        'Wildcard symbols cannot define payout rules.',
      );
    }
    const payoutSymbol = payoutSymbolsByCode.get(rule.symbolMobileCode);
    if (payoutSymbol === undefined) {
      throw new DomainValidationError(
        'incomplete_payout_symbols',
        `Symbol ${rule.symbolMobileCode} has no payout configuration.`,
      );
    }
    if (
      !Number.isSafeInteger(rule.matchLength) ||
      rule.matchLength < payoutSymbol.minimumMatchLength ||
      rule.matchLength > game.columns
    ) {
      throw new DomainValidationError(
        'invalid_match_length',
        `Match length for symbol ${rule.symbolMobileCode} must be between ${payoutSymbol.minimumMatchLength} and ${game.columns}.`,
      );
    }
    requireNonNegativeInteger(
      rule.payoutCredits,
      'invalid_payout',
      'Payout credits',
    );
    const key = `${rule.symbolMobileCode}:${rule.matchLength}`;
    if (keys.has(key)) {
      throw new DomainValidationError(
        'duplicate_payout_rule',
        'Payout rule symbol and match length must be unique.',
      );
    }
    keys.add(key);
  }
}

export function validatePayoutConfiguration(
  rules: readonly PayoutRuleDefinition[],
  payoutSymbols: readonly PayoutSymbolDefinition[],
  game: GameConfig,
): void {
  validatePayoutRules(rules, payoutSymbols, game);
  const payoutSymbolsByCode = new Map(
    payoutSymbols.map((symbol) => [symbol.symbolMobileCode, symbol]),
  );
  const ordinarySymbols = game.symbols.filter((symbol) => !symbol.isWildcard);

  const missingSymbols = ordinarySymbols
    .filter((symbol) => !payoutSymbolsByCode.has(symbol.mobileCode))
    .map((symbol) => symbol.mobileCode);
  if (
    missingSymbols.length > 0 ||
    payoutSymbols.length !== ordinarySymbols.length
  ) {
    throw new DomainValidationError(
      'incomplete_payout_symbols',
      `Missing payout configuration for symbols ${missingSymbols.join(', ')}.`,
    );
  }

  const rulesBySymbol = new Map<number, Map<number, number>>();
  for (const rule of rules) {
    let symbolRules = rulesBySymbol.get(rule.symbolMobileCode);
    if (symbolRules === undefined) {
      symbolRules = new Map<number, number>();
      rulesBySymbol.set(rule.symbolMobileCode, symbolRules);
    }
    symbolRules.set(rule.matchLength, rule.payoutCredits);
  }

  for (const payoutSymbol of payoutSymbols) {
    const symbolRules =
      rulesBySymbol.get(payoutSymbol.symbolMobileCode) ??
      new Map<number, number>();
    let previousPayout: number | undefined;

    for (
      let length = payoutSymbol.minimumMatchLength;
      length <= game.columns;
      length += 1
    ) {
      const payout = symbolRules.get(length);
      if (payout === undefined) {
        throw new DomainValidationError(
          'incomplete_payout_rules',
          `Symbol ${payoutSymbol.symbolMobileCode} is missing payout rule for length ${length}.`,
        );
      }
      if (previousPayout !== undefined && payout <= previousPayout) {
        throw new DomainValidationError(
          'non_increasing_payout',
          `Payout for symbol ${payoutSymbol.symbolMobileCode} must increase with match length.`,
        );
      }
      previousPayout = payout;
    }
  }
}

File: services/worker/src/game_predictor_worker/domain/validation.py
"""Validation at domain boundaries."""

from __future__ import annotations

from collections.abc import Sequence

from game_predictor_worker.domain.contracts import (
    GameConfig,
    PaylineDefinition,
    PayoutRuleDefinition,
    PayoutSymbolDefinition,
    SymbolDefinition,
)
from game_predictor_worker.domain.errors import DomainErrorCode, DomainValidationError
from game_predictor_worker.domain.signature import (
    MAX_SIGNATURE_CELL_WIDTH,
    MAX_SYMBOL_MOBILE_CODE,
)


def _require_non_empty(value: str, code: DomainErrorCode) -> None:
    if not value.strip():
        raise DomainValidationError(code, "Value must not be empty.")


def _require_non_negative_integer(value: int, code: DomainErrorCode, label: str) -> None:
    if isinstance(value, bool) or value < 0:
        raise DomainValidationError(code, f"{label} must be a non-negative integer.")


def validate_board_dimensions(rows: int, columns: int) -> None:
    if isinstance(rows, bool) or isinstance(columns, bool) or rows < 1 or columns < 1:
        raise DomainValidationError(
            DomainErrorCode.INVALID_DIMENSIONS,
            "Board dimensions must be positive integers.",
        )


def _validate_symbols(symbols: Sequence[SymbolDefinition], cell_width: int) -> None:
    if not symbols:
        raise DomainValidationError(
            DomainErrorCode.INVALID_SYMBOL,
            "Game must define at least one symbol.",
        )

    mobile_codes: set[int] = set()
    codes: set[str] = set()
    maximum_code = min(10**cell_width - 1, MAX_SYMBOL_MOBILE_CODE)

    for symbol in symbols:
        _require_non_empty(symbol.code, DomainErrorCode.INVALID_SYMBOL)
        _require_non_empty(symbol.name, DomainErrorCode.INVALID_SYMBOL)
        if (
            isinstance(symbol.mobile_code, bool)
            or symbol.mobile_code < 1
            or symbol.mobile_code > maximum_code
        ):
            raise DomainValidationError(
                DomainErrorCode.INVALID_SYMBOL_CODE,
                f"Symbol mobile code must fit signature width {cell_width}.",
            )
        if isinstance(symbol.display_order, bool) or symbol.display_order < 0:
            raise DomainValidationError(
                DomainErrorCode.INVALID_SYMBOL,
                "Symbol display order must be a non-negative integer.",
            )
        if symbol.mobile_code in mobile_codes:
            raise DomainValidationError(
                DomainErrorCode.DUPLICATE_SYMBOL_MOBILE_CODE,
                f"Duplicate symbol mobile code {symbol.mobile_code}.",
            )
        if symbol.code in codes:
            raise DomainValidationError(
                DomainErrorCode.DUPLICATE_SYMBOL_CODE,
                f"Duplicate symbol code {symbol.code}.",
            )
        mobile_codes.add(symbol.mobile_code)
        codes.add(symbol.code)


def validate_game_config(game: GameConfig) -> None:
    _require_non_empty(game.id, DomainErrorCode.INVALID_GAME)
    _require_non_empty(game.code, DomainErrorCode.INVALID_GAME)
    _require_non_empty(game.name, DomainErrorCode.INVALID_GAME)
    validate_board_dimensions(game.rows, game.columns)
    _require_non_negative_integer(
        game.spin_cost,
        DomainErrorCode.INVALID_SPIN_COST,
        "Spin cost",
    )
    if (
        isinstance(game.signature_cell_width, bool)
        or game.signature_cell_width < 1
        or game.signature_cell_width > MAX_SIGNATURE_CELL_WIDTH
    ):
        raise DomainValidationError(
            DomainErrorCode.INVALID_CELL_WIDTH,
            f"Signature cell width must be between 1 and {MAX_SIGNATURE_CELL_WIDTH}.",
        )
    _validate_symbols(game.symbols, game.signature_cell_width)


def _validate_populated_cell(cell: int, allowed_codes: frozenset[int]) -> None:
    if isinstance(cell, bool) or cell not in allowed_codes:
        raise DomainValidationError(
            DomainErrorCode.INVALID_BOARD_SYMBOL,
            f"Symbol mobile code {cell} does not belong to the game.",
        )


def validate_full_board(cells: Sequence[int], game: GameConfig) -> None:
    validate_game_config(game)
    expected_length = game.rows * game.columns
    if len(cells) != expected_length:
        raise DomainValidationError(
            DomainErrorCode.INVALID_BOARD_LENGTH,
            f"Board contains {len(cells)} cells; expected {expected_length}.",
        )

    allowed_codes = frozenset(symbol.mobile_code for symbol in game.symbols)
    for cell in cells:
        _validate_populated_cell(cell, allowed_codes)


def validate_board_prefix(cells: Sequence[int | None], game: GameConfig) -> None:
    validate_game_config(game)
    expected_length = game.rows * game.columns
    if len(cells) != expected_length:
        raise DomainValidationError(
            DomainErrorCode.INVALID_BOARD_LENGTH,
            f"Board contains {len(cells)} cells; expected {expected_length}.",
        )

    allowed_codes = frozenset(symbol.mobile_code for symbol in game.symbols)
    reached_empty_cell = False
    for cell in cells:
        if cell is None:
            reached_empty_cell = True
            continue
        if reached_empty_cell:
            raise DomainValidationError(
                DomainErrorCode.NON_PREFIX_BOARD,
                "A populated cell cannot occur after an empty prefix cell.",
            )
        _validate_populated_cell(cell, allowed_codes)


def validate_row_path(
    row_path: Sequence[int],
    rows: int,
    columns: int,
) -> None:
    validate_board_dimensions(rows, columns)
    if len(row_path) != columns:
        raise DomainValidationError(
            DomainErrorCode.INVALID_ROW_PATH_LENGTH,
            f"Payline contains {len(row_path)} rows; expected {columns}.",
        )
    if any(
        isinstance(row_index, bool) or row_index < 0 or row_index >= rows for row_index in row_path
    ):
        raise DomainValidationError(
            DomainErrorCode.INVALID_ROW_INDEX,
            f"Every payline row index must be between 0 and {rows - 1}.",
        )


def validate_paylines(
    paylines: Sequence[PaylineDefinition],
    game: GameConfig,
) -> None:
    validate_game_config(game)
    paths: set[tuple[int, ...]] = set()
    ids: set[str] = set()

    for payline in paylines:
        _require_non_empty(payline.id, DomainErrorCode.INVALID_PAYLINE_ID)
        validate_row_path(payline.row_path, game.rows, game.columns)
        if payline.row_path in paths or payline.id in ids:
            raise DomainValidationError(
                DomainErrorCode.DUPLICATE_PAYLINE,
                "Payline id and row path must be unique.",
            )
        paths.add(payline.row_path)
        ids.add(payline.id)


def validate_payout_symbols(
    payout_symbols: Sequence[PayoutSymbolDefinition],
    game: GameConfig,
) -> None:
    validate_game_config(game)
    symbols_by_code = {symbol.mobile_code: symbol for symbol in game.symbols}
    configured_symbols: set[int] = set()

    for payout_symbol in payout_symbols:
        symbol = symbols_by_code.get(payout_symbol.symbol_mobile_code)
        if symbol is None:
            raise DomainValidationError(
                DomainErrorCode.INVALID_BOARD_SYMBOL,
                (f"Payout symbol {payout_symbol.symbol_mobile_code} does not belong to the game."),
            )
        if symbol.is_wildcard:
            raise DomainValidationError(
                DomainErrorCode.WILDCARD_PAYOUT_SYMBOL,
                "Wildcard symbols cannot define payout configuration.",
            )
        if (
            isinstance(payout_symbol.minimum_match_length, bool)
            or payout_symbol.minimum_match_length < 2
            or payout_symbol.minimum_match_length > game.columns
        ):
            raise DomainValidationError(
                DomainErrorCode.INVALID_MINIMUM_MATCH_LENGTH,
                (f"Minimum match length must be between 2 and {game.columns}."),
            )
        if payout_symbol.symbol_mobile_code in configured_symbols:
            raise DomainValidationError(
                DomainErrorCode.DUPLICATE_PAYOUT_SYMBOL,
                (f"Duplicate payout configuration for symbol {payout_symbol.symbol_mobile_code}."),
            )
        configured_symbols.add(payout_symbol.symbol_mobile_code)


def validate_payout_rules(
    rules: Sequence[PayoutRuleDefinition],
    payout_symbols: Sequence[PayoutSymbolDefinition],
    game: GameConfig,
) -> None:
    validate_payout_symbols(payout_symbols, game)
    symbols_by_code = {symbol.mobile_code: symbol for symbol in game.symbols}
    payout_symbols_by_code = {symbol.symbol_mobile_code: symbol for symbol in payout_symbols}
    keys: set[tuple[int, int]] = set()

    for rule in rules:
        symbol = symbols_by_code.get(rule.symbol_mobile_code)
        if symbol is None:
            raise DomainValidationError(
                DomainErrorCode.INVALID_BOARD_SYMBOL,
                f"Payout symbol {rule.symbol_mobile_code} does not belong to the game.",
            )
        if symbol.is_wildcard:
            raise DomainValidationError(
                DomainErrorCode.WILDCARD_PAYOUT_RULE,
                "Wildcard symbols cannot define payout rules.",
            )
        payout_symbol = payout_symbols_by_code.get(rule.symbol_mobile_code)
        if payout_symbol is None:
            raise DomainValidationError(
                DomainErrorCode.INCOMPLETE_PAYOUT_SYMBOLS,
                f"Symbol {rule.symbol_mobile_code} has no payout configuration.",
            )
        if (
            isinstance(rule.match_length, bool)
            or rule.match_length < payout_symbol.minimum_match_length
            or rule.match_length > game.columns
        ):
            raise DomainValidationError(
                DomainErrorCode.INVALID_MATCH_LENGTH,
                (
                    f"Match length for symbol {rule.symbol_mobile_code} must be "
                    f"between {payout_symbol.minimum_match_length} "
                    f"and {game.columns}."
                ),
            )
        _require_non_negative_integer(
            rule.payout_credits,
            DomainErrorCode.INVALID_PAYOUT,
            "Payout credits",
        )
        key = (rule.symbol_mobile_code, rule.match_length)
        if key in keys:
            raise DomainValidationError(
                DomainErrorCode.DUPLICATE_PAYOUT_RULE,
                "Payout rule symbol and match length must be unique.",
            )
        keys.add(key)


def validate_payout_configuration(
    rules: Sequence[PayoutRuleDefinition],
    payout_symbols: Sequence[PayoutSymbolDefinition],
    game: GameConfig,
) -> None:
    """Validate a complete rules matrix ready for payout precomputing."""

    validate_payout_rules(rules, payout_symbols, game)
    payout_symbols_by_code = {
        payout_symbol.symbol_mobile_code: payout_symbol for payout_symbol in payout_symbols
    }
    ordinary_symbols = tuple(symbol for symbol in game.symbols if not symbol.is_wildcard)
    missing_symbols = [
        symbol.mobile_code
        for symbol in ordinary_symbols
        if symbol.mobile_code not in payout_symbols_by_code
    ]
    if missing_symbols or len(payout_symbols) != len(ordinary_symbols):
        raise DomainValidationError(
            DomainErrorCode.INCOMPLETE_PAYOUT_SYMBOLS,
            f"Missing payout configuration for symbols {missing_symbols}.",
        )

    rules_by_symbol: dict[int, dict[int, int]] = {}
    for rule in rules:
        rules_by_symbol.setdefault(rule.symbol_mobile_code, {})[rule.match_length] = (
            rule.payout_credits
        )

    for payout_symbol in payout_symbols:
        expected_lengths = tuple(range(payout_symbol.minimum_match_length, game.columns + 1))
        symbol_rules = rules_by_symbol.get(payout_symbol.symbol_mobile_code, {})
        missing_lengths = [length for length in expected_lengths if length not in symbol_rules]
        if missing_lengths:
            raise DomainValidationError(
                DomainErrorCode.INCOMPLETE_PAYOUT_RULES,
                (
                    f"Symbol {payout_symbol.symbol_mobile_code} is missing payout rules "
                    f"for lengths {missing_lengths}."
                ),
            )

        previous_payout: int | None = None
        for length in expected_lengths:
            payout = symbol_rules[length]
            if previous_payout is not None and payout <= previous_payout:
                raise DomainValidationError(
                    DomainErrorCode.NON_INCREASING_PAYOUT,
                    (
                        f"Payout for symbol {payout_symbol.symbol_mobile_code} "
                        "must increase "
                        f"with match length; length {length} has value {payout}."
                    ),
                )
            previous_payout = payout
